src/server/api/routers/auth.ts

```diff
@@ -62,43 +62,163 @@ export const authRouter = createTRPCRouter({
         profileImageURL: string | null;
-      };
-      let participants: Participant[] = [];
-      if (input.filters[0] == "gender") {
-        participants = await ctx.db.query.user.findMany({
-          where: and(
-            eq(user.role, "participant"),
-            eq(user.role, "participant"),
-          ),
-          columns: {
-            aka: true,
-            bio: true,
-            email: true,
-            firstName: true,
-            gender: true,
-            role: true,
-            lastName: true,
-            profileImageURL: true,
-          },
-        });
+        interests: unknown;
+      }[] = [];
+      if (input.gender != "" && input.interests.length != 0) {
+        select = await ctx.db
+          .select({
+            aka: user.aka,
+            bio: user.bio,
+            email: user.email,
+            firstName: user.firstName,
+            gender: user.gender,
+            role: user.role,
+            lastName: user.lastName,
+            dateOfBirth: user.dateOfBirth,
+            profileImageURL: user.profileImageURL,
+            interests: sql`GROUP_CONCAT(${hostInterest.interest}) AS interests`,
+          })
+          .from(hostUser)
+          .where(
+            and(
+              inArray(
+                hostUser.userID,
+                ctx.db
+                  .select({ userID: hostUser.userID })
+                  .from(hostUser)
+                  .innerJoin(hostInterest, eq(hostInterest.userID, user.id))
+                  .where(and(inArray(hostInterest.interest, input.interests)))
+                  .groupBy(hostUser.userID)
+                  .having(
+                    sql`count(${hostUser.userID}) >= ${input.interests.length}`,
+                  ),
+              ),
+              gte(user.dateOfBirth, minDate),
+              lte(user.dateOfBirth, maxDate),
+              gte(hostUser.avgRating, input.rating),
+              eq(user.gender, input.gender),
+            ),
+          )
+          .innerJoin(user, eq(user.id, hostUser.userID))
+          .innerJoin(hostInterest, eq(hostInterest.userID, hostUser.userID))
+          .groupBy(hostUser.userID);
+        return select;
+      }
+      if (input.gender != "" && input.interests.length == 0) {
+        select = await ctx.db
+          .select({
+            aka: user.aka,
+            bio: user.bio,
+            email: user.email,
+            firstName: user.firstName,
+            gender: user.gender,
+            role: user.role,
+            lastName: user.lastName,
+            dateOfBirth: user.dateOfBirth,
+            profileImageURL: user.profileImageURL,
+            interests: sql`GROUP_CONCAT(${hostInterest.interest}) AS interests`,
+          })
+          .from(hostUser)
+          .where(
+            and(
+              gte(user.dateOfBirth, minDate),
+              lte(user.dateOfBirth, maxDate),
+              gte(hostUser.avgRating, input.rating),
+              eq(user.gender, input.gender),
```

Comment:
ปัญหาคือเราไป fixed ว่าใน and ต้องมีอะไรบ้าง
จริงๆ ที่เราควรทำคือ สร้าง array conditions มา
แล้วก็ append ของ condition เข้าไปถ้าเขาส่งมา
แล้วค่อย spread มันแบบ `and(conditions...)` ถ้าทำแบบนี้มันก็ไม่ต้องมาเขียนยาวๆซ้ำๆแบบนี้